### ISSUE TITLE
Add stereoscopic rendering to Godot 4

### ISSUE THREAD

**Issue by BastiaanOlij** [godotengine/godot-proposals#2648]
### Describe the project you are working on
Bringing XR capabilities back to Godot 4

### Describe the problem or limitation you are having in your project
The current approach for rendering stereoscopic images for purposes such as XR in Godot 3 is not working in the new Vulkan renderer. The approach is also an older one where some preparation is done up front but then each eye is rendered one after the other and sent to the XR compositor individually. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
The idea for Godot 4 is to create a more modern stereoscopic rendering pipeline where both eyes are rendered in parallel.
With stereoscopic rendering both eyes generally look in the same direction the eyes will basically see the same objects, there may be small differences but as long as objects only visible by one eye aren't culled out the gains by not duplicating state changes will far outweigh the overhead of attempting to render that object to both eyes.

Vulkan even has automatic support for this in the [multiview](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VK_KHR_multiview.html) extension. If available all render buffer images can simply have two layers and all draw calls are automatically duplicated.
It should be available in Vulkan 1.1 on wards. We'll need to look into if it is likely XR applications would run on Vulkan 1.0 devices where this extension is optional and whether a fallback mode should be created.

This proposal assumes we will be able to rely on multiview, if a fallback is required that will be added afterwards.
Also for now we're looking purely into the use case of stereoscopic rendering with 2 render surfaces (left + right eye), multiview can be used to render any number of surfaces with widely varying cameras though this will have an impact on how objects will be culled.

An example of a non-XR application of multiview is in higher end driving and flight simulators where often monitors are placed to the side of a player to render the views for when the player looks out the side windows. 

We should keep this in mind when building our solution as to not close that door but this is not the use case we're going for and those situations are probably just as efficiently implemented simply using multiple viewports for multiple monitors.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
Currently the difference between rendering mono or stereoscopic is controlled by the xr setting on a viewport. If XR is enabled this enables stereoscopic rendering IF the active XRinterface requests this.

For now we'll keep that behavior the same but what we will change is how this is handled in the render server when the viewport is processed. 

` RendererViewport::draw_viewports` will be simplified and simply parse on whether xr mode is active and whether it is set to stereoscopic. 

There is a question here whether we should disable 2D rendering at this point as 2D rendering in the way it is implemented in Godot doesn't have any real use in this scenario. We could do a stereoscopic projection at this point but it would require serious changes to the 2D rendering pipeline with very little benefit over the developer just using another viewport and a viewport texture on a quad. 

In `RendererViewport::_draw_3d` where we currently call `render_camera` we start deviating. Right now this function takes all the properties of one camera as parameters, the projection matrix, the transform, etc. 
This needs to change to an array of cameras. One for mono, two for stereo.

`render_camera` will run through the render server to render the 3D view, first going through the culling process, then calling the actual rendering logic held within the active renderer.

Frustum culling will require only simple changes as we can cull based on a combined frustum as Godot 3 currently does. We may pick up objects only visible by one eye but as mentioned before the performance benefits outweigh the few false positives here.

Occlusion culling on the other hand needs more thought. You can test this out yourself really simple. Stand by a wall and peek into a room (common thing to do in FPS games). One eye will see everything in the room, the other eye just sees the wall. You can use a center position but still risk culling everything the eye that's peeking sees and this will be incredibly obvious to the player. VR is wicked that way. @JFonS any ideas how to work around this problem?

The render loop itself can remain mostly unchanged other then getting an array of camera data instead of a single camera. The rest of the changes will mostly be done in the render device itself where our color and depth textures will require two layers and where the multiview feature needs to be enabled. 

Our SceneData UBO will also need to get the projection matrix, camera matrix and MVP matrix duplicated. The multiview extension introduces a new variable called `ViewIndex` which tells our vertex and fragment shaders whether it is being called for the left eye or right eye and thus which matrices to use.  We'll need to make sure the `ViewIndex` exists when multiview is not being used as well but just has the value of 0 so we use the correct matrices. 

Finally we need a way to send the end result back to the XRinterface so it can be passed on to the compositor. We'll make this work with the build in native mobile interface first (it also works on desktop).

In line with this, at this stage we are not looking at the current external texture implementation used by Oculus and OpenXR. The goal is to get stereoscopic working again as a first step, for now we can use the old workaround copying the end result into the buffers supplied by those SDKs and then look into improving this. 

### If this enhancement will not be used often, can it be worked around with a few lines of script?
No.

### Is there a reason why this should be core and not an add-on in the asset library?
This is a core feature

**Comment by BastiaanOlij** [comment]
Just to highlight the peek problem:

![image](https://user-images.githubusercontent.com/1945449/115986855-acd2d180-a5f5-11eb-8c3a-faa90687acf1.png)

![image](https://user-images.githubusercontent.com/1945449/115986865-b9572a00-a5f5-11eb-9625-b357219d8946.png)

**Comment by reduz** [comment]
Occlusion culling could probably just work by interleaving rays from the and right eye horizontally, since afterwards it checks for vicinity to decide the maximum depth seen.

**Comment by BastiaanOlij** [comment]
> 
> 
> Occlusion culling could probably just work by interleaving rays from the and right eye horizontally, since afterwards it checks for vicinity to decide the maximum depth seen.

Sounds like a really good idea, is the occlusion culling using a form of raytracing then? I haven't read up on how the solution jfons is implementing works internally.

**Comment by JFonS** [comment]
Regarding occlusion culling I see three ways to handle this:

* **Single occlusion buffer + increased Znear**: For general scenes we can probably get away with just using a single occlusion buffer but increasing its Znear value. Objects closer to the camera have a higher chance of producing different occlusions per-eye, so getting rid of close objects should give mostly correct results. It's a hacky solution and still has corner cases, but it might be worth trying.

* **Single occlusion buffer with interleaved rays**: As @reduz proposed, having a checkerboard pattern of rays from both eyes should work pretty well. My only issue with this approach is it's essentially the same as having two separate lower resolution buffers with the advantage of checking them both at once, but needs a bit more work if we want to mix the rays but still keep them coherent.

* **Two occlusion buffers**: Since testing is the fastest part of the process, we can try just having a separate (lower resolution) buffer for each eye and testing each object against both buffers. This would give the best results, and I'm pretty sure it's fast enough for desktops, but would need to be tested on mobile.

Bottom line, I would not worry too much about OC for now. Once we have merged both the occlusion culling PR and the stereoscopic rendering, I can poke around and see which of the proposed ideas works for us.

**Comment by BastiaanOlij** [comment]
@JFonS agreed, I'm flagging this because it's a potential issue with OC, until we have stereoscopic working again in Godot 4 on an actual headset with a good test environment it'll be hard to judge which strategy works best.

**Comment by BastiaanOlij** [comment]
While there are still a number of effects that need to be improved, the core of stereoscopic functionality is now fully merged in both mobile and clustered renderers so this can be closed :)
